src/control/lib/netdetect/netdetect.go

```diff
@@ -271,30 +292,38 @@ func getHwlocDeviceNames(deviceScanCfg DeviceScan) ([]string, error) {
 
 // initDeviceScan initializes the hwloc library and initializes DeviceScan entries
 // that are frequently used for hwloc queries
-func initDeviceScan() (DeviceScan, error) {
+func initDeviceScan(ndc *NetDetectContext) (DeviceScan, error) {
 	var deviceScanCfg DeviceScan
+	var err error
 
-	topology, err := initLib()
-	if err != nil {
-		log.Debugf("Error from initLib %v", err)
-		return deviceScanCfg,
-			errors.New("unable to initialize hwloc library")
+	if ndc == nil {
```

Comment:
Not a big deal to leave it is as-is, but I have a suggestion. Right now there are a lot of places where we're checking if ndc == nil. Makes me think perhaps the shared portion could be extracted as a helper function, called "populateDeviceScan" or something like that, and accepts a pointer to a DeviceScan with the topology already set. Then the cases with a NetDetectContext could call that directly, or call a very thin wrapper that creates the DeviceScan from the NetDetectContext.
For the other cases, we could keep initDeviceScan with no param passed in, which would create the topology and a DeviceScan, pass it in to be populated, and if an error was returned, free the topology in one place. Again, just a suggestion, but it might make things a little cleaner here. Otherwise looks good to me.

Reply:
I'm actually still looking at this change.  I am not sure if it makes sense to do that in this PR or not.  I do like the idea of calling initDeviceScan at netdetect.Init() because that would be cleaner.  I will think about that some more and see what it would take to do that.

Reply:
The change was made.